Potential bug in timer due date checker scheduling

**Summary**

> Try to answer the following as best as possible

- How often does the test fail? [once](https://github.com/camunda/zeebe/actions/runs/8520495239/attempts/1)
- Does it block your work? no
- Do we suspect that it is a real failure? not sure.

**Failures**

> Outline known failure cases, e.g. a failed assertion and its stacktrace obtained from Jenkins

<details><summary>Example assertion failure</summary>
<pre>
 shouldTriggerOnlyTimerStartEvent (Flaky Test): io.camunda.zeebe.engine.processing.timer.TimerStartEventTest
java.lang.AssertionError: 

Expecting ArrayList:
  []
to contain only:
  ["timer_start"]
but could not find the following element(s):
  ["timer_start"]

	at io.camunda.zeebe.engine.processing.timer.TimerStartEventTest.shouldTriggerOnlyTimerStartEvent(TimerStartEventTest.java:883)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
</pre>
</details>

**Hypotheses**

> List any hypotheses if you have one; can be ommitted
No clue!
**Logs**

> If possible, provide more context here, e.g. standard output logs, link to build, etc.

<details><summary>Logs</summary>
<pre>
https://productionresultssa2.blob.core.windows.net/actions-results/0ae2319b-b6c7-4386-ad29-60d7cae419b8/workflow-job-run-b8245a8d-c8fe-58cd-5361-3fdfc2c073c3/logs/job/job-logs.txt?rsct=text%2Fplain&se=2024-04-02T09%3A54%3A54Z&sig=PJkLcZ0vN2NDFsF46Y5K%2FrvoQJI8KncbGDktLQJPwtA%3D&sp=r&spr=https&sr=b&st=2024-04-02T09%3A44%3A49Z&sv=2021-12-02
</pre>
</details>

--- a/zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/scheduled/DueDateChecker.java
+++ b/zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/scheduled/DueDateChecker.java
@@ -7,6 +7,8 @@
  */
 package io.camunda.zeebe.engine.processing.scheduled;
 
+import io.camunda.zeebe.engine.processing.scheduled.DueDateChecker.NextExecution.None;
+import io.camunda.zeebe.engine.processing.scheduled.DueDateChecker.NextExecution.Scheduled;
 import io.camunda.zeebe.scheduler.clock.ActorClock;
 import io.camunda.zeebe.stream.api.ReadonlyStreamProcessorContext;
 import io.camunda.zeebe.stream.api.StreamProcessorLifecycleAware;
@@ -46,7 +48,7 @@ public final class DueDateChecker implements StreamProcessorLifecycleAware {
    * Keeps track of the next execution of the checker. Value can be null if there is no scheduled
    * execution known.
    */
-  private final AtomicReference<NextExecution> nextExecution = new AtomicReference<>(null);
+  private final AtomicReference<NextExecution> nextExecution = new AtomicReference<>(new None());
 
   /**
    * @param timerResolution The resolution in ms for the timer
@@ -63,14 +65,14 @@ public final class DueDateChecker implements StreamProcessorLifecycleAware {
   }
 
   TaskResult execute(final TaskResultBuilder taskResultBuilder) {
-    // There is a benign edge case where we are not supposed to set nextExecution to null here.
+    // There is a benign edge case where we are not supposed to set nextExecution to None here.
     // If this execution was supposed to be cancelled because an earlier execution was scheduled
-    // instead, nextExecution would hold that earlier execution. We still overwrite it with null and
+    // instead, nextExecution would hold that earlier execution. We still overwrite it with None and
     // thus forget that we planned an execution. The next time something is scheduled, we will
-    // observe the null value and thus decide to schedule something new without cancelling what's
+    // observe the None value and thus decide to schedule something new without cancelling what's
     // already scheduled. While we try to avoid this, we can't prevent it entirely anyway because
     // cancellation of scheduled executions is best-effort and does not reliably prevent execution.
-    nextExecution.set(null);
+    nextExecution.set(new None());
 
     final long nextDueDate = visitor.apply(taskResultBuilder);
 
@@ -109,22 +111,21 @@ public final class DueDateChecker implements StreamProcessorLifecycleAware {
     final var replacedExecution =
         AtomicUtil.replace(
             nextExecution,
-            currentlyScheduled -> {
+            currentlyPlanned -> {
               final var now = ActorClock.currentTimeMillis();
-              final long scheduleFor =
-                  dueDate == -1 ? now : now + Math.max(dueDate - now, timerResolution);
-              if (currentlyScheduled == null
-                  || currentlyScheduled.scheduledFor() - scheduleFor > timerResolution) {
+              final long scheduleFor = now + Math.max(dueDate - now, timerResolution);
+              if (!(currentlyPlanned instanceof final Scheduled currentlyScheduled)
+                  || (currentlyScheduled.scheduledFor() - scheduleFor > timerResolution)) {
                 final var delay = Duration.ofMillis(scheduleFor - now);
                 final var task = scheduleService.runDelayed(delay, this::execute);
-                return Optional.of(new NextExecution(scheduleFor, task));
+                return Optional.of(new Scheduled(scheduleFor, task));
               }
               return Optional.empty();
             },
-            newlyScheduled -> newlyScheduled.task().cancel());
+            NextExecution::cancel);
 
     if (replacedExecution != null) {
-      replacedExecution.task().cancel();
+      replacedExecution.cancel();
     }
   }
 
@@ -163,14 +164,6 @@ public final class DueDateChecker implements StreamProcessorLifecycleAware {
   }
 
   /**
-   * Keeps track of the next execution of the checker.
-   *
-   * @param scheduledFor The deadline in ms for when this execution is scheduled.
-   * @param task The scheduled task for the next execution, can be used for canceling the task.
-   */
-  private record NextExecution(long scheduledFor, ScheduledTask task) {}
-
-  /**
    * Abstracts over async and sync scheduling methods of {@link
    * io.camunda.zeebe.stream.api.scheduling.ProcessingScheduleService}.
    */
@@ -184,5 +177,30 @@ public final class DueDateChecker implements StreamProcessorLifecycleAware {
      * Task)}
      */
     ScheduledTask runDelayed(final Duration delay, final Task task);
+  }
+
+  interface NextExecution {
+    void cancel();
+
+    /** Sentinel value to signal that nothing is scheduled. */
+    record None() implements NextExecution {
+
+      @Override
+      public void cancel() {}
+    }
+
+    /**
+     * Keeps track of the next execution of the checker.
+     *
+     * @param scheduledFor The deadline in ms for when this execution is scheduled.
+     * @param task The scheduled task for the next execution, can be used for canceling the task.
+     */
+    record Scheduled(long scheduledFor, ScheduledTask task) implements NextExecution {
+
+      @Override
+      public void cancel() {
+        task.cancel();
+      }
+    }
   }
 }

--- a/zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/scheduled/DueDateCheckerTest.java
+++ b/zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/scheduled/DueDateCheckerTest.java
@@ -37,7 +37,7 @@ public class DueDateCheckerTest {
 
     when(mockContext.getScheduleService()).thenReturn(mockScheduleService);
     dueDateChecker.onRecovered(mockContext);
-    verify(mockScheduleService).runDelayed(eq(Duration.ZERO), any(Task.class));
+    verify(mockScheduleService).runDelayed(eq(Duration.ofMillis(timerResolution)), any(Task.class));
     dueDateChecker.execute(mock(TaskResultBuilder.class));
     Mockito.clearInvocations(mockScheduleService);
 
@@ -63,7 +63,7 @@ public class DueDateCheckerTest {
 
     when(mockContext.getScheduleService()).thenReturn(mockScheduleService);
     dueDateChecker.onRecovered(mockContext);
-    verify(mockScheduleService).runDelayed(eq(Duration.ZERO), any(Task.class));
+    verify(mockScheduleService).runDelayed(eq(Duration.ofMillis(timerResolution)), any(Task.class));
     dueDateChecker.execute(mock(TaskResultBuilder.class));
     Mockito.clearInvocations(mockScheduleService);
 
@@ -93,7 +93,7 @@ public class DueDateCheckerTest {
 
     when(mockContext.getScheduleService()).thenReturn(mockScheduleService);
     dueDateChecker.onRecovered(mockContext);
-    verify(mockScheduleService).runDelayed(eq(Duration.ZERO), any(Task.class));
+    verify(mockScheduleService).runDelayed(eq(Duration.ofMillis(timerResolution)), any(Task.class));
     Mockito.clearInvocations(mockScheduleService);
 
     // when
@@ -119,7 +119,7 @@ public class DueDateCheckerTest {
 
     when(mockContext.getScheduleService()).thenReturn(mockScheduleService);
     dueDateChecker.onRecovered(mockContext);
-    verify(mockScheduleService).runDelayed(eq(Duration.ZERO), any(Task.class));
+    verify(mockScheduleService).runDelayed(eq(Duration.ofMillis(timerResolution)), any(Task.class));
     Mockito.clearInvocations(mockScheduleService);
 
     dueDateChecker.execute(mock(TaskResultBuilder.class));

--- a/zeebe/scheduler/src/main/java/io/camunda/zeebe/scheduler/clock/ControlledActorClock.java
+++ b/zeebe/scheduler/src/main/java/io/camunda/zeebe/scheduler/clock/ControlledActorClock.java
@@ -14,6 +14,7 @@ import java.time.Instant;
 public final class ControlledActorClock implements ActorClock {
   private volatile long currentTime;
   private volatile long currentOffset;
+  private volatile long updatedTime;
 
   public ControlledActorClock() {
     reset();
@@ -34,6 +35,7 @@ public final class ControlledActorClock implements ActorClock {
   public void reset() {
     currentTime = -1;
     currentOffset = 0;
+    updatedTime = System.currentTimeMillis();
   }
 
   public Instant getCurrentTime() {
@@ -54,16 +56,17 @@ public final class ControlledActorClock implements ActorClock {
 
   @Override
   public boolean update() {
+    if (usesPointInTime()) {
+      updatedTime = currentTime;
+    } else {
+      updatedTime = System.currentTimeMillis() + currentOffset;
+    }
     return true;
   }
 
   @Override
   public long getTimeMillis() {
-    if (usesPointInTime()) {
-      return currentTime;
-    } else {
-      return System.currentTimeMillis() + currentOffset;
-    }
+    return updatedTime;
   }
 
   @Override

--- a/zeebe/util/src/main/java/io/camunda/zeebe/util/AtomicUtil.java
+++ b/zeebe/util/src/main/java/io/camunda/zeebe/util/AtomicUtil.java
@@ -25,8 +25,8 @@ public final class AtomicUtil {
    * replaced.
    *
    * <p>If the value of the atomic reference has been replaced by another thread in the meantime,
-   * the rollback function is called with the value provided by the replacer function, and the
-   * replacer function is called again.
+   * the replacer function is called again. Additionally, the rollback function is applied to any
+   * non-empty value returned by the previous replacer function call.
    *
    * @param ref The atomic reference that holds the value to replace
    * @param replacer The replacer function used to provide a new value to the atomic reference; if
@@ -40,20 +40,17 @@ public final class AtomicUtil {
       final AtomicReference<T> ref,
       final Function<T, Optional<T>> replacer,
       final Consumer<T> rollback) {
-    T currentVal;
+    T currentVal = null;
     T newVal = null;
     do {
-      if (newVal != null) {
+      if (newVal != null && newVal != currentVal) {
         // another thread appears to have replaced the value in the meantime
         rollback.accept(newVal);
       }
       currentVal = ref.get();
       final var result = replacer.apply(currentVal);
-      if (result.isEmpty()) {
-        return null;
-      }
-      newVal = result.get();
+      newVal = result.orElse(currentVal);
     } while (!ref.compareAndSet(currentVal, newVal));
-    return currentVal;
+    return currentVal == newVal ? null : currentVal;
   }
 }

